[BUG] Orchestrion pin v1.6.0

Version of orchestrion
v1.6.0

Describe what happened:

When running orchestrion pin, the command fails with a panic due to the instrument directory being missing inside the orchestrion module path.

The error points to:

```
panic: resolving github.com/DataDog/dd-trace-go/orchestrion/all/v2 version in
"/Users/<user>/go/pkg/mod/github.com/!data!dog/orchestrion@v1.6.0/instrument": 
getting GOMOD: running ["go" "env" "GOMOD"]: chdir
/Users/<user>/go/pkg/mod/github.com/!data!dog/orchestrion@v1.6.0/instrument: 
no such file or directory
```


The instrument directory is not present in the v1.6.0 module, causing orchestrion pin to fail even though the rest of the setup (tool.go imports and go.mod dependencies) works correctly.

Describe what you expected:

orchestrion pin should complete successfully — updating go.mod and generating orchestrion.tool.go as intended — without requiring manual workarounds or path adjustments.

Steps to reproduce the issue:

Install orchestrion v1.6.0:

`go install github.com/DataDog/orchestrion@v1.6.0`


Initialize a Go project and run:

`orchestrion pin`


The command panics due to a missing /instrument directory inside the orchestrion module.

The expected directory structure:

/Users/<user>/go/pkg/mod/github.com/!data!dog/orchestrion@v1.6.0/instrument


is missing in the published module.

Workaround:

You can still successfully build and run the project using:

`go build -o <output> ./cmd/<your-app>`


because the dependency references (orchestrion.tool.go + github.com/DataDog/dd-trace-go/orchestrion/all/v2) are correct.
However, the orchestrion pin command remains unusable in v1.6.0.

Additional environment details:

Go version: go1.22 (reproducible also with go1.21)

macOS (ARM and Intel tested)

Project layout: standard Go module workspace

--- a/internal/ensure/requiredintegrations.go
+++ b/internal/ensure/requiredintegrations.go
@@ -196,23 +196,41 @@ var (
 
 func fetchShippedVersions() map[string]string {
 	initOnce.Do(func() {
-		versions := make(map[string]string)
 		_, thisFile, _, _ := runtime.Caller(0)
-		// The version of dd-trace-go that shipped with the current version of orchestrion.
-		// We use this to determine if we need to upgrade dd-trace-go when pinning.
 		orchestrionRoot := filepath.Join(thisFile, "..", "..", "..")
-		ver, err := resolveDependencyVersion(orchestrionRoot, integrations.DatadogTracerV2)
-		if err != nil {
-			panic(fmt.Errorf("resolving %s version in %q: %w", integrations.DatadogTracerV2, orchestrionRoot, err))
-		}
-		versions[integrations.DatadogTracerV2] = ver
-		instrumentationRoot := filepath.Join(orchestrionRoot, "instrument")
-		ver, err = resolveDependencyVersion(instrumentationRoot, integrations.DatadogTracerV2All)
-		if err != nil {
-			panic(fmt.Errorf("resolving %s version in %q: %w", integrations.DatadogTracerV2All, instrumentationRoot, err))
-		}
-		versions[integrations.DatadogTracerV2All] = ver
+		versions := loadShippedVersions(orchestrionRoot)
 		orchestrionShippedVersions.Store(&versions)
 	})
 	return *orchestrionShippedVersions.Load()
+}
+
+func loadShippedVersions(orchestrionRoot string) map[string]string {
+	versions := make(map[string]string)
+
+	// The version of dd-trace-go that shipped with the current version of orchestrion.
+	// We use this to determine if we need to upgrade dd-trace-go when pinning.
+	ver, err := resolveDependencyVersion(orchestrionRoot, integrations.DatadogTracerV2)
+	if err != nil {
+		panic(fmt.Errorf("resolving %s version in %q: %w", integrations.DatadogTracerV2, orchestrionRoot, err))
+	}
+	versions[integrations.DatadogTracerV2] = ver
+
+	// Default to using the same version as DatadogTracerV2 for DatadogTracerV2All.
+	// This serves as a fallback when the instrument directory (which is a separate submodule)
+	// is not accessible, such as when orchestrion is used as a module dependency from the
+	// Go module cache where nested submodules may not be present.
+	versions[integrations.DatadogTracerV2All] = versions[integrations.DatadogTracerV2]
+
+	// If the instrument directory exists, override with the actual version from its go.mod
+	instrumentationRoot := filepath.Join(orchestrionRoot, "instrument")
+	if _, err := os.Stat(instrumentationRoot); err != nil {
+		return versions
+	}
+	ver, err = resolveDependencyVersion(instrumentationRoot, integrations.DatadogTracerV2All)
+	if err != nil {
+		panic(fmt.Errorf("resolving %s version in %q: %w", integrations.DatadogTracerV2All, instrumentationRoot, err))
+	}
+	versions[integrations.DatadogTracerV2All] = ver
+
+	return versions
 }

--- a/internal/ensure/requiredintegrations_test.go
+++ b/internal/ensure/requiredintegrations_test.go
@@ -10,6 +10,7 @@ import (
 	"errors"
 	"fmt"
 	"os"
+	"path/filepath"
 	"testing"
 
 	"github.com/DataDog/orchestrion/internal/gomod"
@@ -410,6 +411,24 @@ func TestFetchShippedVersions(t *testing.T) {
 				assert.True(t, semver.IsValid(version), "version %q for %q should be valid semver", version, key)
 			}
 		}
+	})
+
+	t.Run("fallback-when-instrument-missing", func(t *testing.T) {
+		// Create temp directory without instrument/ subdirectory
+		tempDir := t.TempDir()
+		testGoMod := filepath.Join(tempDir, "go.mod")
+		require.NoError(t, os.WriteFile(testGoMod, []byte(`module test
+require github.com/DataDog/dd-trace-go/v2 v2.3.0
+`), 0644))
+
+		versions := loadShippedVersions(tempDir)
+
+		v2Version := versions["github.com/DataDog/dd-trace-go/v2"]
+		allVersion := versions["github.com/DataDog/dd-trace-go/orchestrion/all/v2"]
+
+		require.NotEmpty(t, v2Version)
+		require.NotEmpty(t, allVersion)
+		assert.Equal(t, v2Version, allVersion, "both should use same version when instrument/ is missing")
 	})
 }
 

